%% Generated by Sphinx.
\def\sphinxdocclass{report}
\documentclass[letterpaper,11pt,english]{sphinxmanual}
\ifdefined\pdfpxdimen
   \let\sphinxpxdimen\pdfpxdimen\else\newdimen\sphinxpxdimen
\fi \sphinxpxdimen=.75bp\relax

\usepackage[utf8]{inputenc}
\ifdefined\DeclareUnicodeCharacter
 \ifdefined\DeclareUnicodeCharacterAsOptional\else
  \DeclareUnicodeCharacter{00A0}{\nobreakspace}
\fi\fi
\usepackage{cmap}
\usepackage[T1]{fontenc}
\usepackage{amsmath,amssymb,amstext}
\usepackage{babel}
\usepackage{times}
\usepackage[Bjarne]{fncychap}
\usepackage[dontkeepoldnames]{sphinx}

\usepackage{geometry}

% Include hyperref last.
\usepackage{hyperref}
% Fix anchor placement for figures with captions.
\usepackage{hypcap}% it must be loaded after hyperref.
% Set up styles of URL: it should be placed after hyperref.
\urlstyle{same}
\addto\captionsenglish{\renewcommand{\contentsname}{Contents:}}

\addto\captionsenglish{\renewcommand{\figurename}{Fig.}}
\addto\captionsenglish{\renewcommand{\tablename}{Table}}
\addto\captionsenglish{\renewcommand{\literalblockname}{Listing}}

\addto\extrasenglish{\def\pageautorefname{page}}

\setcounter{tocdepth}{1}



\title{MVFtools Documentation}
\date{Jun 24, 2017}
\release{2017-06-24}
\author{James B. Pease, Ben K. Rosenzweig}
\newcommand{\sphinxlogo}{\sphinxincludegraphics{logo.png}\par}
\renewcommand{\releasename}{Release}
\makeindex

\begin{document}

\maketitle
\sphinxtableofcontents
\phantomsection\label{\detokenize{index::doc}}



\chapter{Getting Started}
\label{\detokenize{intro:intro}}\label{\detokenize{intro:welcome-to-mvftools-s-documentation}}\label{\detokenize{intro::doc}}\label{\detokenize{intro:getting-started}}

\section{What is MVFtools?}
\label{\detokenize{intro:what-is-mvftools}}
Multisample Variant Format (MVF), is designed for compact storage and efficient analysis of multi-genome and multi-transcriptome datasets.  The programs provided in MVFtools support this format, both with conversion utilities, filtering and transformation programs, and data analysis and visualization modules.  MVF format is designed specifically for biological data analysis, since sequence data is encoded based on the information content at a particular aligned sequence site.  This contextual encoding allows for rapid computation of phylogenetic and population genetic analyses, and small file sizes that enable data sharing and distribution.


\section{How do I cite this ?}
\label{\detokenize{intro:how-do-i-cite-this}}
Pease JB and BK Rosenzweig. 2016. “Encoding Data Using Biological Principles: the Multisample Variant Format for Phylogenomics and Population Genomics” \sphinxstyleemphasis{IEEE/ACM Transactions on Computational Biology and Bioinformatics}. In press. \sphinxurl{http://www.dx.doi.org/10.1109/tcbb.2015.2509997}

Please also include the URL \textless{}\sphinxurl{https://www.github.com/jbpease/mvftools}\textgreater{} in your methods section where the program is referenced.


\section{Installation}
\label{\detokenize{intro:installation}}
No installation is required, mvftools scripts should work as long as Python3 is installed.  The repository can be cloned or downloaded as a .zip file from GitHub.
\begin{description}
\item[{::}] \leavevmode
git clone \sphinxurl{https://www.github.com/jbpease/mvftools}

\end{description}

Alternatively, you can download MVftools as a .zip file from the github page.


\subsection{Requirements}
\label{\detokenize{intro:requirements}}\begin{itemize}
\item {} 
Python 3.x (2.7 should also work, but 3.x recommended) \sphinxurl{https://www.python.org/downloads/}

\item {} 
RAxML 8.x (7.x should also work, but 8.x recommended) \sphinxurl{https://sco.h-its.org/exelixis/web/software/raxml/index.html}

\end{itemize}


\subsection{Additional Requirements for Some Modules:}
\label{\detokenize{intro:additional-requirements-for-some-modules}}\begin{itemize}
\item {} 
Scipy: (\sphinxurl{http://www.scipy.org/})

\item {} 
Biopython 1.6+: (\sphinxurl{http://www.biopython.org/}),

\item {} 
Numpy (\sphinxurl{http://www.numpy.org/}),

\item {} 
RAxML (\sphinxurl{http://sco.h-its.org/exelixis/web/software/raxml/index.html/})

\end{itemize}


\section{Preparing your data}
\label{\detokenize{intro:preparing-your-data}}

\subsection{Sequence Alignment}
\label{\detokenize{intro:sequence-alignment}}
MVF files can be created from VCF, FASTA, and MAF files using the \sphinxcode{vcf2mvf.py}, \sphinxcode{fasta2mvf.py}, or \sphinxcode{maf2mvf.py} programs respectively.  Once converted to MVF format, analyses and manipulations can be carried out using the rest of the tools in MVFtools.


\section{Basic usage examples}
\label{\detokenize{intro:basic-usage-examples}}
\sphinxstylestrong{Case \#1: Generate phylogenies from 100kb windows using a VCF data}:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{python3} \PYG{n}{vcf2mvf}\PYG{o}{.}\PYG{n}{py} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{vcf} \PYG{n}{DATA}\PYG{o}{.}\PYG{n}{vcf} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{mvf} \PYG{n}{DATA}\PYG{o}{.}\PYG{n}{mvf}
\PYG{n}{python3} \PYG{n}{mvf\PYGZus{}window\PYGZus{}tree}\PYG{o}{.}\PYG{n}{py} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{mvf} \PYG{n}{DATA}\PYG{o}{.}\PYG{n}{mvf} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{out} \PYG{n}{WINDOWTREES}\PYG{o}{.}\PYG{n}{txt} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{windowsize} \PYG{l+m+mi}{100000}
\end{sphinxVerbatim}

\sphinxstylestrong{Case \#2: Convert a large FASTA file, then generate window-based counts for DFOIL/D-statistic introgression testing from the first five samples}:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{python3} \PYG{n}{fasta2mvf}\PYG{o}{.}\PYG{n}{py} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{fasta} \PYG{n}{DATA}\PYG{o}{.}\PYG{n}{fasta} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{mvf} \PYG{n}{DATA}\PYG{o}{.}\PYG{n}{mvf}
\PYG{n}{python3} \PYG{n}{mvf\PYGZus{}analyze\PYGZus{}dna}\PYG{o}{.}\PYG{n}{py} \PYG{n}{PatternCount} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{mvf} \PYG{n}{DATA}\PYG{o}{.}\PYG{n}{mvf} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{out} \PYG{n}{PATTERNS}\PYG{o}{.}\PYG{n}{txt} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{windowsize} \PYG{l+m+mi}{100000} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{samples} \PYG{l+m+mi}{0} \PYG{l+m+mi}{1} \PYG{l+m+mi}{2} \PYG{l+m+mi}{3} \PYG{l+m+mi}{4}
\end{sphinxVerbatim}

The file is now ready to use as an input file for with dfoil (\sphinxurl{http://www.github.com/jbpease/dfoil}).


\chapter{MVF Format Specification (version 1.2)}
\label{\detokenize{mvf_spec:mvf-format-specification-version-1-2}}\label{\detokenize{mvf_spec::doc}}

\section{Version History}
\label{\detokenize{mvf_spec:version-history}}

\subsection{v1.1.1}
\label{\detokenize{mvf_spec:v1-1-1}}
Codons and Proteins accommodated


\subsection{v1.2}
\label{\detokenize{mvf_spec:v1-2}}
Dot masking, multi-line header, adoption of “X” in place of “N” for nucleotides, support for non-reference aligned sequences.


\section{MVF General Notes and Usage}
\label{\detokenize{mvf_spec:mvf-general-notes-and-usage}}

\subsection{General Features}
\label{\detokenize{mvf_spec:general-features}}
MVF is primarily intended for site-wise analyses in phylogenomics and population genomics. MVF is formatted to contain one aligned site per line, but contains only allelic information, therefore MVF most closely mimics VCF files in formatting, but resembles MAF format in informational content,  Additionally, MVF uses special formatting to lower file sizes and speed up filtering and analysis.  MVF can readily be adapted from other common sequence formats including VCF, FSATA, and MAF.  MVF is also designed to be able to accommodate readily store other information for phylogenomic projects, including tree topologies and sample metadata.


\subsection{Native Gzip reat/write}
\label{\detokenize{mvf_spec:native-gzip-reat-write}}
MVF is designed to work natively with GZIP compression and uses a formatting that attempts to strike a balance between fast filtering, easy visual inspection, while using character patterns that create a good Gzip compression ratio. As long as any input or output file path ends with exactly “.gz”, all MVF scripts will natively read/write to gzip-compressed files.


\subsection{General Notes on Filtering}
\label{\detokenize{mvf_spec:general-notes-on-filtering}}
MVF was specifically designed as a “vertical” format for rapid filtering of \sphinxstyleemphasis{sites} in large-scale phylogenomic analyses. (rather than being “horizontal” to visually show alignment) Therefore, the following should be noted to take advantage of MVF formatting for rapid filtering (i.e. with grep/zgrep).
\begin{itemize}
\item {} 
\sphinxcode{\#} is present iff. the line is in the header

\item {} 
\sphinxcode{@} is present iff. the position is non-reference

\item {} 
\sphinxcode{X} is present in the allele string iff. the positon has ambiguity data

\item {} 
\sphinxcode{\#:} can quickly filter by chromosome

\item {} 
\sphinxcode{:\#} can quickly filter by coordinate numbers

\item {} 
Allele strings with one or two characters have full sample coverage (no gaps)

\item {} 
Allele strings with \sphinxcode{@{[}any{]}+} have coverage=1, \sphinxcode{{[}not@{]}{[}any{]}+} have coverage=2

\item {} 
One or two-character allele strings, or notation with \sphinxcode{{[}any{]}+} CANNOT contain homoplasy or synapomorphy (by definition).

\end{itemize}


\section{Header Specification}
\label{\detokenize{mvf_spec:header-specification}}
All header lines begin with one or more \sphinxcode{\#} and contain single-space separated fields.


\subsection{MVF declaration line}
\label{\detokenize{mvf_spec:mvf-declaration-line}}
First header line always starts with \sphinxcode{\#\#mvf}, followed by required metadata fields:
\begin{itemize}
\item {} 
version=1.2

\item {} 
mvftype={[}dna, protein, codon{]}

\end{itemize}

and optionally:
\begin{itemize}
\item {} 
an arbitrary number of metadata fields in key=value format (‘mvftype’ and ‘version’ not allowed as key)

\end{itemize}


\subsection{Sample information}
\label{\detokenize{mvf_spec:sample-information}}
Sample information (columns) header lines are specified by:
* line starts with \sphinxcode{\#s} (“s” for sample) with no leading spaces
* LABEL (must be unique, no spaces)
* an arbitrary number of metadata fields in key=value format (‘label’ not allowed as key)

The first entry should be the reference sequence (if aligned to reference) or can be any sequence in the case of non-reference-aligned de novo alignment).


\subsection{Contig information}
\label{\detokenize{mvf_spec:contig-information}}
Contig information header lines are specified by:
* line starts with \sphinxcode{\#c} (“c” for contig)
* CONTIG\_ID (must be unique, alpha-numeric strong recommended, must not contain \sphinxcode{*:;,@!+} or spaces)
* \sphinxcode{label={[}NAME{]}} (recommended by not required to be unique, no spaces allowed)
* \sphinxcode{len={[}LENGTH{]}} (integer \textgreater{} 0, or zero for unknown)
* \sphinxcode{ref={[}0/1{]}}, indicates if contig is reference-based (=1) or not (=0)
* an arbitrary number of metadata fields in key=value format (“label”, “len”, and “ref” not allowed as key)


\subsection{Tree information}
\label{\detokenize{mvf_spec:tree-information}}
Tree information may (optionally) be specified in header lines by:
* line starts with \sphinxcode{\#t} (“t” for tree/topology)
* \sphinxcode{TREE\_ID={[}\#\#\#{]}{}` (must be unique, alpha-numeric)
* {}`{}`TOPOLOGY={[}tree\_String{]}} in Newick/Phylip/parenthetical format (must end with ‘;’)
* an arbitrary number of metadata fields in key=value format
To take full advantage of MVF tree storage, use the same sample labels as in the \sphinxcode{\#s} header lines


\subsection{Notes}
\label{\detokenize{mvf_spec:notes}}
General project notes may (optionally) be specified in the header lines by:
* line starts with \sphinxcode{\#n} (“n” for notes)
* Text is unstructured and is not necessarily formatted as metadata


\subsection{Example Header}
\label{\detokenize{mvf_spec:example-header}}\begin{description}
\item[{::}] \leavevmode
\#\#mvf version=1.2 mvftype={[}MVFTYPE{]}
\#s SAMPLE0 meta0=somevalue meta1=0 …
\#s SAMPLE1 meta0=somethingele meta1=1 …
\#s SAMPLE2 meta0=somesome meta1=0 …
…
\#c 0 name=CONTIG0 length=100 ref=1 meta0=somevalue …
\#c 1 name=CONTIG1 length=200 ref=0 meta0=someother …
…
\#t 0 ((SAMPLE0,SAMPLE1),SAMPLE2); model=GTRGAMMA software=RAxML
\#t 1 ((SAMPLE2,SAMPLE0),SAMPLE1); model=GTRGAMMA software=RAxML partition=chrom1
…
\#n Notes on this project.

\end{description}


\section{Entry Specification}
\label{\detokenize{mvf_spec:entry-specification}}
\begin{sphinxadmonition}{note}{Note:}
all examples show an MVF entry with REF and four samples
\end{sphinxadmonition}

Entries are structured as two space-separated columns:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
ID:POSITION   ALLELES [ALLELES ALLELES ...]

* {}`{}`ID:POSITION{}`{}` = chromosomal id matching the first element of a contig in the {}`{}`\PYGZsh{}c{}`{}` header element
* {}`{}`POSITION{}`{}` = 1\PYGZhy{}based position on the contig with matching {}`{}`CONTIG\PYGZus{}ID{}`{}`
* {}`{}`ALLELES{}`{}` = one or more records of alleles at reference\PYGZhy{}based location specified by {}`{}`ID:POSITION{}`{}` and matching the formatting below
\end{sphinxVerbatim}


\subsection{For mvftype=codon}
\label{\detokenize{mvf_spec:for-mvftype-codon}}\begin{itemize}
\item {} 
Allele columns are \sphinxcode{PROTEIN DNA1 DNA2 DNA3} where the three DNA columns represent three codon positions in collated form

\item {} 
Position is the position of the lowest numbered codon position (regardless of transcript strand) and \sphinxcode{DNA1/2/3} codon columns are given in order to match the protein (again regardless of transcript orientation)

\end{itemize}


\subsection{Allele formatting}
\label{\detokenize{mvf_spec:allele-formatting}}
\begin{sphinxadmonition}{note}{Note:}
all examples show an MVF entry with five samples.
\end{sphinxadmonition}

For reference-anchored contigs, the first allele is assumed to be the “reference” allele by default. Each entry must either (1) contain the same number of characters as sample labels specified in the header or (2) use one of the special cases in the section below.

\sphinxcode{ATCTG} =  (REF is ‘A’ samples 1\&3 are ‘T’, sample 2 is ‘C’, sample 4 is ‘G’)


\subsection{Special cases}
\label{\detokenize{mvf_spec:special-cases}}
\begin{sphinxadmonition}{note}{Note:}
all examples show an MVF entry with five samples
\end{sphinxadmonition}


\subsection{Invariant sites}
\label{\detokenize{mvf_spec:invariant-sites}}
When all alleles are both present (non-gap) and all the same, this is represented by a single base.
\begin{quote}

\sphinxcode{A = AAAAA}
\end{quote}


\subsection{Monoallelic non-reference samples}
\label{\detokenize{mvf_spec:monoallelic-non-reference-samples}}
When all alleles in the samples (non-REF) are the same but differ from REF, this is represented by two bases.
\begin{quote}

\sphinxcode{AT = ATTTT}
\sphinxcode{Aa = Aaaaa}
\end{quote}


\subsection{Single-variant sites}
\label{\detokenize{mvf_spec:single-variant-sites}}
When only one of the samples varies from the others, this is specified as:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{p}{[}\PYG{n}{reference\PYGZus{}base}\PYG{p}{,} \PYG{n}{majority\PYGZus{}base}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{+}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{unique\PYGZus{}base}\PYG{p}{,} \PYG{n}{unique\PYGZus{}position}\PYG{p}{]}
\end{sphinxVerbatim}

This is useful shorthand for both sites with one a single base that differs and samples with only one sample represented.  When the site only has coverage via one sample (i.e. all other bases are empty, the ‘-‘ is omitted from the second position.
\begin{quote}

\sphinxcode{AC+T2 = ACTCC}
\sphinxcode{AA+C2 = AACAA}
\sphinxcode{-+A2  = -{-}A-{-}}
\sphinxcode{A+A2  = A-A-{-}}
\sphinxcode{A+a2  = A-a-{-}}
\sphinxcode{A+C2  = A-C-{-}}
\end{quote}


\subsection{Non-reference aligned sites}
\label{\detokenize{mvf_spec:non-reference-aligned-sites}}
Added in MVF v.1.2, this facilitates using MVF for non-reference aligned sequences (e.g. aligned sets of orthologs from de novo assembled transcripts). These non-reference-anchored alignments can comprise the entire MVF file or be included in addition to reference-aligned contigs. Non-reference-contigs in their header entry should include the keyword “nonref” (see Section 1.3). Contigs labels and coordinates are labelled the same as reference-based entries. To denote that the sequence is non-reference and not simply a deletion in the reference, the character “@” should be the first character of the alignment.  In the case an entirely non-reference MVF, all contigs can be labelled as “nonref,” but one sequence should be chosen as the reference for the purposes of the allele
string.  When this sequence is not present, \sphinxcode{@} is still used.
\begin{quote}

\sphinxcode{@AATT   = -AATT}
\sphinxcode{@A+T3   = -A-T-}
\sphinxcode{@-+A3   = -{-}-A-}
\end{quote}


\section{Character encoding}
\label{\detokenize{mvf_spec:character-encoding}}

\subsection{Nucleotide Notation}
\label{\detokenize{mvf_spec:nucleotide-notation}}\begin{itemize}
\item {} 
Standard IUPAC nucleotide codes are used: \sphinxcode{ACGT}, and \sphinxcode{U} for uracil in RNA

\item {} 
Standard IUPAC bialleic ambiguity codes \sphinxcode{KMRSWY} are used also.

\item {} 
Current MVF formatting does NOT allow triallelic ambiguity codes (\sphinxcode{BDHV}), which are converted to ambiguous (\sphinxcode{X}) instead.

\item {} 
Current MVF formatting does NOT recognize rare symbols (\sphinxcode{ISOX}, or \sphinxcode{Phi})

\item {} 
Ambiguous nucleotide is denoted by \sphinxcode{X} instead of standard \sphinxcode{N}

\end{itemize}


\subsection{Amino Acid Notation}
\label{\detokenize{mvf_spec:amino-acid-notation}}\begin{itemize}
\item {} 
Standard IUPAC amino acid codes are used: \sphinxcode{ACDEFGHIKLMNPQRSTVWY}

\item {} 
Standard stop codon symbol \sphinxcode{*} is used

\item {} 
Currently the ambiguous/rare symbols are not recognized (\sphinxcode{BZ})

\end{itemize}


\subsection{Use of \sphinxstyleliteralintitle{X} for ambiguous nucleotides and amino acids}
\label{\detokenize{mvf_spec:use-of-x-for-ambiguous-nucleotides-and-amino-acids}}
In standard notation, “\sphinxcode{N}” is used for an ambiguous nucleotide, which could be any of A/C/G/T.
However, in amino acid notation \sphinxcode{N} stands for “Asparagine” and is a valid character, while \sphinxcode{X} is used for an ambiguous amino acid.
MVF v1.2 adopts \sphinxcode{X} as unified ambiguity character for both nucleotides and proteins for MVF files for two purposes:
1. To creates a unified ambiguity character for MVF codon files for faster processing
2. To allow fast filtering of ambiguous lines
Also note that while ‘X’ in expanded IUPAC notation refers to ‘xanthosine,’ MVF currently does not support rare nucleotides.
.. note:: In all conversion utilities that export from MVF format to another file format conversion to the standard “N”/”X” for ambiguous nucleotides/amino acids should ALWAYS be implemented.


\chapter{Examples of the same data in MVF Format and other formats}
\label{\detokenize{mvf_example::doc}}\label{\detokenize{mvf_example:examples-of-the-same-data-in-mvf-format-and-other-formats}}

\section{MVF Format}
\label{\detokenize{mvf_example:mvf-format}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}mvf sourceformat=fasta version=1.2 mvftype=dna ncol=5}
\PYG{c+c1}{\PYGZsh{}s Hsapiens}
\PYG{c+c1}{\PYGZsh{}s Ptroglodytes}
\PYG{c+c1}{\PYGZsh{}s Ppaniscus}
\PYG{c+c1}{\PYGZsh{}s Ggorilla}
\PYG{c+c1}{\PYGZsh{}s Mmusculus}
\PYG{c+c1}{\PYGZsh{}c 1 Chromosome1 length=248956422}
\PYG{c+c1}{\PYGZsh{}n Note: This is an example file showing data formatting}
\PYG{l+m+mi}{1}\PYG{p}{:}\PYG{l+m+mi}{100} \PYG{n}{A}
\PYG{l+m+mi}{1}\PYG{p}{:}\PYG{l+m+mi}{101} \PYG{n}{A}
\PYG{l+m+mi}{1}\PYG{p}{:}\PYG{l+m+mi}{102} \PYG{n}{A}
\PYG{l+m+mi}{1}\PYG{p}{:}\PYG{l+m+mi}{103} \PYG{n}{T}
\PYG{l+m+mi}{1}\PYG{p}{:}\PYG{l+m+mi}{104} \PYG{n}{TT}\PYG{o}{+}\PYG{n}{C4}
\PYG{l+m+mi}{1}\PYG{p}{:}\PYG{l+m+mi}{105} \PYG{n}{GC}
\PYG{l+m+mi}{1}\PYG{p}{:}\PYG{l+m+mi}{106} \PYG{n}{A}\PYG{o}{+}\PYG{n}{A4}
\PYG{l+m+mi}{1}\PYG{p}{:}\PYG{l+m+mi}{107} \PYG{n}{AATTA}
\PYG{l+m+mi}{1}\PYG{p}{:}\PYG{l+m+mi}{108} \PYG{n}{AC}\PYG{o}{+}\PYG{n}{G4}
\end{sphinxVerbatim}


\section{FASTA Format}
\label{\detokenize{mvf_example:fasta-format}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{o}{\PYGZgt{}}\PYG{n}{Hsapiens} \PYG{n}{gi}\PYG{p}{:}\PYG{l+m+mi}{1234} \PYG{n}{geneid}\PYG{p}{:}\PYG{n}{GeneOfInterest} \PYG{n}{chrom}\PYG{p}{:}\PYG{l+m+mi}{1} \PYG{n}{start}\PYG{p}{:}\PYG{l+m+mi}{100} \PYG{n}{end}\PYG{p}{:}\PYG{l+m+mi}{108}
\PYG{n}{AAATTGAAA}

\PYG{o}{\PYGZgt{}}\PYG{n}{Ptroglodytes} \PYG{n}{geneid}\PYG{p}{:}\PYG{n}{GeneOfInterest}
\PYG{n}{AAATTC}\PYG{o}{\PYGZhy{}}\PYG{n}{AC}

\PYG{o}{\PYGZgt{}}\PYG{n}{Ppaniscus} \PYG{n}{geneid}\PYG{p}{:}\PYG{n}{GeneOfInterest}
\PYG{n}{AAATTC}\PYG{o}{\PYGZhy{}}\PYG{n}{TC}

\PYG{o}{\PYGZgt{}}\PYG{n}{Ggorilla} \PYG{n}{geneid}\PYG{p}{:}\PYG{n}{GeneOfInterest}
\PYG{n}{AAATTC}\PYG{o}{\PYGZhy{}}\PYG{n}{TC}

\PYG{o}{\PYGZgt{}}\PYG{n}{Mmusculus} \PYG{n}{geneid}\PYG{p}{:}\PYG{n}{GeneOfInterest}
\PYG{n}{AAATCCAAG}
\end{sphinxVerbatim}


\section{VCF Format}
\label{\detokenize{mvf_example:vcf-format}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}fileformat=VCFv4.1}
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}samtoolsVersion=0.1.19\PYGZhy{}44428cd}
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}reference=hg19.fa}
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}contig=\PYGZlt{}ID=Chromosome1,length=248956422\PYGZgt{}}
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}INFO=\PYGZlt{}ID=DP,Number=1,Type=Integer,Description=\PYGZdq{}Raw read depth\PYGZdq{}\PYGZgt{}}
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}INFO=\PYGZlt{}ID=DP4,Number=4,Type=Integer,Description=\PYGZdq{}\PYGZsh{} high\PYGZhy{}quality ref\PYGZhy{}forward bases, ref\PYGZhy{}reverse, alt\PYGZhy{}forward and alt\PYGZhy{}reverse bases\PYGZdq{}\PYGZgt{}}
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}INFO=\PYGZlt{}ID=MQ,Number=1,Type=Integer,Description=\PYGZdq{}Root\PYGZhy{}mean\PYGZhy{}square mapping quality of covering reads\PYGZdq{}\PYGZgt{}}
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}INFO=\PYGZlt{}ID=FQ,Number=1,Type=Float,Description=\PYGZdq{}Phred probability of all samples being the same\PYGZdq{}\PYGZgt{}}
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}INFO=\PYGZlt{}ID=AF1,Number=1,Type=Float,Description=\PYGZdq{}Max\PYGZhy{}likelihood estimate of the first ALT allele frequency (assuming HWE)\PYGZdq{}\PYGZgt{}}
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}INFO=\PYGZlt{}ID=AC1,Number=1,Type=Float,Description=\PYGZdq{}Max\PYGZhy{}likelihood estimate of the first ALT allele count (no HWE assumption)\PYGZdq{}\PYGZgt{}}
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}INFO=\PYGZlt{}ID=AN,Number=1,Type=Integer,Description=\PYGZdq{}Total number of alleles in called genotypes\PYGZdq{}\PYGZgt{}}
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}INFO=\PYGZlt{}ID=IS,Number=2,Type=Float,Description=\PYGZdq{}Maximum number of reads supporting an indel and fraction of indel reads\PYGZdq{}\PYGZgt{}}
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}INFO=\PYGZlt{}ID=AC,Number=A,Type=Integer,Description=\PYGZdq{}Allele count in genotypes for each ALT allele, in the same order as listed\PYGZdq{}\PYGZgt{}}
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}INFO=\PYGZlt{}ID=G3,Number=3,Type=Float,Description=\PYGZdq{}ML estimate of genotype frequencies\PYGZdq{}\PYGZgt{}}
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}INFO=\PYGZlt{}ID=HWE,Number=1,Type=Float,Description=\PYGZdq{}Chi\PYGZca{}2 based HWE test P\PYGZhy{}value based on G3\PYGZdq{}\PYGZgt{}}
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}INFO=\PYGZlt{}ID=CLR,Number=1,Type=Integer,Description=\PYGZdq{}Log ratio of genotype likelihoods with and without the constraint\PYGZdq{}\PYGZgt{}}
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}INFO=\PYGZlt{}ID=UGT,Number=1,Type=String,Description=\PYGZdq{}The most probable unconstrained genotype configuration in the trio\PYGZdq{}\PYGZgt{}}
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}INFO=\PYGZlt{}ID=CGT,Number=1,Type=String,Description=\PYGZdq{}The most probable constrained genotype configuration in the trio\PYGZdq{}\PYGZgt{}}
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}INFO=\PYGZlt{}ID=PV4,Number=4,Type=Float,Description=\PYGZdq{}P\PYGZhy{}values for strand bias, baseQ bias, mapQ bias and tail distance bias\PYGZdq{}\PYGZgt{}}
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}INFO=\PYGZlt{}ID=INDEL,Number=0,Type=Flag,Description=\PYGZdq{}Indicates that the variant is an INDEL.\PYGZdq{}\PYGZgt{}}
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}INFO=\PYGZlt{}ID=PC2,Number=2,Type=Integer,Description=\PYGZdq{}Phred probability of the nonRef allele frequency in group1 samples being larger (,smaller) than in group2.\PYGZdq{}\PYGZgt{}}
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}INFO=\PYGZlt{}ID=PCHI2,Number=1,Type=Float,Description=\PYGZdq{}Posterior weighted chi\PYGZca{}2 P\PYGZhy{}value for testing the association between group1 and group2 samples.\PYGZdq{}\PYGZgt{}}
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}INFO=\PYGZlt{}ID=QCHI2,Number=1,Type=Integer,Description=\PYGZdq{}Phred scaled PCHI2.\PYGZdq{}\PYGZgt{}}
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}INFO=\PYGZlt{}ID=PR,Number=1,Type=Integer,Description=\PYGZdq{}\PYGZsh{} permutations yielding a smaller PCHI2.\PYGZdq{}\PYGZgt{}}
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}INFO=\PYGZlt{}ID=QBD,Number=1,Type=Float,Description=\PYGZdq{}Quality by Depth: QUAL/\PYGZsh{}reads\PYGZdq{}\PYGZgt{}}
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}INFO=\PYGZlt{}ID=RPB,Number=1,Type=Float,Description=\PYGZdq{}Read Position Bias\PYGZdq{}\PYGZgt{}}
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}INFO=\PYGZlt{}ID=MDV,Number=1,Type=Integer,Description=\PYGZdq{}Maximum number of high\PYGZhy{}quality nonRef reads in samples\PYGZdq{}\PYGZgt{}}
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}INFO=\PYGZlt{}ID=VDB,Number=1,Type=Float,Description=\PYGZdq{}Variant Distance Bias (v2) for filtering splice\PYGZhy{}site artefacts in RNA\PYGZhy{}seq data. Note: this version may be broken.\PYGZdq{}\PYGZgt{}}
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}FORMAT=\PYGZlt{}ID=GT,Number=1,Type=String,Description=\PYGZdq{}Genotype\PYGZdq{}\PYGZgt{}}
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}FORMAT=\PYGZlt{}ID=GQ,Number=1,Type=Integer,Description=\PYGZdq{}Genotype Quality\PYGZdq{}\PYGZgt{}}
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}FORMAT=\PYGZlt{}ID=GL,Number=3,Type=Float,Description=\PYGZdq{}Likelihoods for RR,RA,AA genotypes (R=ref,A=alt)\PYGZdq{}\PYGZgt{}}
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}FORMAT=\PYGZlt{}ID=DP,Number=1,Type=Integer,Description=\PYGZdq{}\PYGZsh{} high\PYGZhy{}quality bases\PYGZdq{}\PYGZgt{}}
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}FORMAT=\PYGZlt{}ID=DV,Number=1,Type=Integer,Description=\PYGZdq{}\PYGZsh{} high\PYGZhy{}quality non\PYGZhy{}reference bases\PYGZdq{}\PYGZgt{}}
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}FORMAT=\PYGZlt{}ID=SP,Number=1,Type=Integer,Description=\PYGZdq{}Phred\PYGZhy{}scaled strand bias P\PYGZhy{}value\PYGZdq{}\PYGZgt{}}
\PYG{c+c1}{\PYGZsh{}\PYGZsh{}FORMAT=\PYGZlt{}ID=PL,Number=G,Type=Integer,Description=\PYGZdq{}List of Phred\PYGZhy{}scaled genotype likelihoods\PYGZdq{}\PYGZgt{}}
\PYG{c+c1}{\PYGZsh{}CHROM        POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  Ptroglodytes Ppaniscus  Ggorilla        Mmusculus}
\PYG{n}{ch01}  \PYG{l+m+mi}{100}     \PYG{o}{.}       \PYG{n}{A}       \PYG{o}{.}       \PYG{l+m+mi}{30}      \PYG{o}{.}       \PYG{n}{DP}\PYG{o}{=}\PYG{l+m+mi}{5}\PYG{p}{;}\PYG{n}{AF1}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{;}\PYG{n}{AC1}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{;}\PYG{n}{DP4}\PYG{o}{=}\PYG{l+m+mi}{5}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{;}\PYG{n}{MQ}\PYG{o}{=}\PYG{l+m+mi}{20}\PYG{p}{;}\PYG{n}{FQ}\PYG{o}{=}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{23.4}     \PYG{n}{PL}\PYG{p}{:}\PYG{n}{DP}   \PYG{l+m+mi}{0}\PYG{o}{/}\PYG{l+m+mi}{0}\PYG{p}{:}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{40}\PYG{p}{:}\PYG{l+m+mi}{2}\PYG{p}{:}\PYG{l+m+mi}{4}  \PYG{l+m+mi}{0}\PYG{o}{/}\PYG{l+m+mi}{0}\PYG{p}{:}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{40}\PYG{p}{:}\PYG{l+m+mi}{2}\PYG{p}{:}\PYG{l+m+mi}{4}  \PYG{l+m+mi}{0}\PYG{o}{/}\PYG{l+m+mi}{0}\PYG{p}{:}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{40}\PYG{p}{:}\PYG{l+m+mi}{2}\PYG{p}{:}\PYG{l+m+mi}{4}  \PYG{l+m+mi}{0}\PYG{o}{/}\PYG{l+m+mi}{0}\PYG{p}{:}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{40}\PYG{p}{:}\PYG{l+m+mi}{2}\PYG{p}{:}\PYG{l+m+mi}{4}
\PYG{n}{ch01}  \PYG{l+m+mi}{101}     \PYG{o}{.}       \PYG{n}{A}       \PYG{o}{.}       \PYG{l+m+mi}{30}      \PYG{o}{.}       \PYG{n}{DP}\PYG{o}{=}\PYG{l+m+mi}{5}\PYG{p}{;}\PYG{n}{AF1}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{;}\PYG{n}{AC1}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{;}\PYG{n}{DP4}\PYG{o}{=}\PYG{l+m+mi}{5}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{;}\PYG{n}{MQ}\PYG{o}{=}\PYG{l+m+mi}{20}\PYG{p}{;}\PYG{n}{FQ}\PYG{o}{=}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{23.4}     \PYG{n}{PL}\PYG{p}{:}\PYG{n}{DP}   \PYG{l+m+mi}{0}\PYG{o}{/}\PYG{l+m+mi}{0}\PYG{p}{:}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{40}\PYG{p}{:}\PYG{l+m+mi}{2}\PYG{p}{:}\PYG{l+m+mi}{4}  \PYG{l+m+mi}{0}\PYG{o}{/}\PYG{l+m+mi}{0}\PYG{p}{:}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{40}\PYG{p}{:}\PYG{l+m+mi}{2}\PYG{p}{:}\PYG{l+m+mi}{4}  \PYG{l+m+mi}{0}\PYG{o}{/}\PYG{l+m+mi}{0}\PYG{p}{:}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{40}\PYG{p}{:}\PYG{l+m+mi}{2}\PYG{p}{:}\PYG{l+m+mi}{4}  \PYG{l+m+mi}{0}\PYG{o}{/}\PYG{l+m+mi}{0}\PYG{p}{:}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{40}\PYG{p}{:}\PYG{l+m+mi}{2}\PYG{p}{:}\PYG{l+m+mi}{4}
\PYG{n}{ch01}  \PYG{l+m+mi}{102}     \PYG{o}{.}       \PYG{n}{A}       \PYG{o}{.}       \PYG{l+m+mi}{30}      \PYG{o}{.}       \PYG{n}{DP}\PYG{o}{=}\PYG{l+m+mi}{5}\PYG{p}{;}\PYG{n}{AF1}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{;}\PYG{n}{AC1}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{;}\PYG{n}{DP4}\PYG{o}{=}\PYG{l+m+mi}{5}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{;}\PYG{n}{MQ}\PYG{o}{=}\PYG{l+m+mi}{20}\PYG{p}{;}\PYG{n}{FQ}\PYG{o}{=}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{23.4}     \PYG{n}{PL}\PYG{p}{:}\PYG{n}{DP}   \PYG{l+m+mi}{0}\PYG{o}{/}\PYG{l+m+mi}{0}\PYG{p}{:}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{40}\PYG{p}{:}\PYG{l+m+mi}{2}\PYG{p}{:}\PYG{l+m+mi}{4}  \PYG{l+m+mi}{0}\PYG{o}{/}\PYG{l+m+mi}{0}\PYG{p}{:}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{40}\PYG{p}{:}\PYG{l+m+mi}{2}\PYG{p}{:}\PYG{l+m+mi}{4}  \PYG{l+m+mi}{0}\PYG{o}{/}\PYG{l+m+mi}{0}\PYG{p}{:}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{40}\PYG{p}{:}\PYG{l+m+mi}{2}\PYG{p}{:}\PYG{l+m+mi}{4}  \PYG{l+m+mi}{0}\PYG{o}{/}\PYG{l+m+mi}{0}\PYG{p}{:}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{40}\PYG{p}{:}\PYG{l+m+mi}{2}\PYG{p}{:}\PYG{l+m+mi}{4}
\PYG{n}{ch01}  \PYG{l+m+mi}{103}     \PYG{o}{.}       \PYG{n}{T}       \PYG{o}{.}       \PYG{l+m+mi}{32}      \PYG{o}{.}       \PYG{n}{DP}\PYG{o}{=}\PYG{l+m+mi}{5}\PYG{p}{;}\PYG{n}{AF1}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{;}\PYG{n}{AC1}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{;}\PYG{n}{DP4}\PYG{o}{=}\PYG{l+m+mi}{5}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{;}\PYG{n}{MQ}\PYG{o}{=}\PYG{l+m+mi}{20}\PYG{p}{;}\PYG{n}{FQ}\PYG{o}{=}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{23.4}     \PYG{n}{PL}\PYG{p}{:}\PYG{n}{DP}   \PYG{l+m+mi}{0}\PYG{o}{/}\PYG{l+m+mi}{0}\PYG{p}{:}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{40}\PYG{p}{:}\PYG{l+m+mi}{2}\PYG{p}{:}\PYG{l+m+mi}{4}  \PYG{l+m+mi}{0}\PYG{o}{/}\PYG{l+m+mi}{0}\PYG{p}{:}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{40}\PYG{p}{:}\PYG{l+m+mi}{2}\PYG{p}{:}\PYG{l+m+mi}{4}  \PYG{l+m+mi}{0}\PYG{o}{/}\PYG{l+m+mi}{0}\PYG{p}{:}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{40}\PYG{p}{:}\PYG{l+m+mi}{2}\PYG{p}{:}\PYG{l+m+mi}{4}  \PYG{l+m+mi}{0}\PYG{o}{/}\PYG{l+m+mi}{0}\PYG{p}{:}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{40}\PYG{p}{:}\PYG{l+m+mi}{2}\PYG{p}{:}\PYG{l+m+mi}{4}
\PYG{n}{ch01}  \PYG{l+m+mi}{104}     \PYG{o}{.}       \PYG{n}{T}       \PYG{n}{C}       \PYG{l+m+mf}{7.61}    \PYG{o}{.}       \PYG{n}{DP}\PYG{o}{=}\PYG{l+m+mi}{2}\PYG{p}{;}\PYG{n}{VDB}\PYG{o}{=}\PYG{l+m+mf}{6.720000e\PYGZhy{}02}\PYG{p}{;}\PYG{n}{AF1}\PYG{o}{=}\PYG{l+m+mi}{1}\PYG{p}{;}\PYG{n}{AC1}\PYG{o}{=}\PYG{l+m+mi}{58}\PYG{p}{;}\PYG{n}{DP4}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{;}\PYG{n}{MQ}\PYG{o}{=}\PYG{l+m+mi}{20}\PYG{p}{;}\PYG{n}{FQ}\PYG{o}{=}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{23.8}   \PYG{n}{GT}\PYG{p}{:}\PYG{n}{PL}\PYG{p}{:}\PYG{n}{DP}\PYG{p}{:}\PYG{n}{GQ}     \PYG{l+m+mi}{0}\PYG{o}{/}\PYG{l+m+mi}{0}\PYG{p}{:}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{40}\PYG{p}{:}\PYG{l+m+mi}{2}\PYG{p}{:}\PYG{l+m+mi}{4}  \PYG{l+m+mi}{0}\PYG{o}{/}\PYG{l+m+mi}{0}\PYG{p}{:}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{40}\PYG{p}{:}\PYG{l+m+mi}{2}\PYG{p}{:}\PYG{l+m+mi}{4}  \PYG{l+m+mi}{0}\PYG{o}{/}\PYG{l+m+mi}{0}\PYG{p}{:}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{40}\PYG{p}{:}\PYG{l+m+mi}{2}\PYG{p}{:}\PYG{l+m+mi}{4}  \PYG{l+m+mi}{1}\PYG{o}{/}\PYG{l+m+mi}{1}\PYG{p}{:}\PYG{l+m+mi}{38}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{:}\PYG{l+m+mi}{2}\PYG{p}{:}\PYG{l+m+mi}{4}
\PYG{n}{ch01}  \PYG{l+m+mi}{105}     \PYG{o}{.}       \PYG{n}{G}       \PYG{n}{C}       \PYG{l+m+mf}{32.1}    \PYG{o}{.}       \PYG{n}{DP}\PYG{o}{=}\PYG{l+m+mi}{5}\PYG{p}{;}\PYG{n}{AF1}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{;}\PYG{n}{AC1}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{;}\PYG{n}{DP4}\PYG{o}{=}\PYG{l+m+mi}{5}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{;}\PYG{n}{MQ}\PYG{o}{=}\PYG{l+m+mi}{20}\PYG{p}{;}\PYG{n}{FQ}\PYG{o}{=}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{23.4}     \PYG{n}{PL}\PYG{p}{:}\PYG{n}{DP}   \PYG{l+m+mi}{0}\PYG{o}{/}\PYG{l+m+mi}{0}\PYG{p}{:}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{40}\PYG{p}{:}\PYG{l+m+mi}{2}\PYG{p}{:}\PYG{l+m+mi}{4}  \PYG{l+m+mi}{0}\PYG{o}{/}\PYG{l+m+mi}{0}\PYG{p}{:}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{40}\PYG{p}{:}\PYG{l+m+mi}{2}\PYG{p}{:}\PYG{l+m+mi}{4}  \PYG{l+m+mi}{0}\PYG{o}{/}\PYG{l+m+mi}{0}\PYG{p}{:}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{40}\PYG{p}{:}\PYG{l+m+mi}{2}\PYG{p}{:}\PYG{l+m+mi}{4}  \PYG{l+m+mi}{1}\PYG{o}{/}\PYG{l+m+mi}{1}\PYG{p}{:}\PYG{l+m+mi}{38}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{:}\PYG{l+m+mi}{2}\PYG{p}{:}\PYG{l+m+mi}{4}
\PYG{n}{ch01}  \PYG{l+m+mi}{106}     \PYG{o}{.}       \PYG{n}{A}       \PYG{o}{.}       \PYG{l+m+mi}{30}      \PYG{o}{.}       \PYG{n}{DP}\PYG{o}{=}\PYG{l+m+mi}{5}\PYG{p}{;}\PYG{n}{AF1}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{;}\PYG{n}{AC1}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{;}\PYG{n}{DP4}\PYG{o}{=}\PYG{l+m+mi}{5}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{;}\PYG{n}{MQ}\PYG{o}{=}\PYG{l+m+mi}{20}\PYG{p}{;}\PYG{n}{FQ}\PYG{o}{=}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{23.4}     \PYG{n}{PL}\PYG{p}{:}\PYG{n}{DP}   \PYG{l+m+mi}{0}\PYG{p}{:}\PYG{l+m+mi}{0}     \PYG{l+m+mi}{0}\PYG{p}{:}\PYG{l+m+mi}{0}     \PYG{l+m+mi}{0}\PYG{p}{:}\PYG{l+m+mi}{0}     \PYG{l+m+mi}{0}\PYG{o}{/}\PYG{l+m+mi}{0}\PYG{p}{:}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{40}\PYG{p}{:}\PYG{l+m+mi}{2}\PYG{p}{:}\PYG{l+m+mi}{4}
\PYG{n}{ch01}  \PYG{l+m+mi}{107}     \PYG{o}{.}       \PYG{n}{A}       \PYG{n}{T}       \PYG{l+m+mf}{24.4}    \PYG{o}{.}       \PYG{n}{DP}\PYG{o}{=}\PYG{l+m+mi}{5}\PYG{p}{;}\PYG{n}{AF1}\PYG{o}{=}\PYG{l+m+mi}{1}\PYG{p}{;}\PYG{n}{AC1}\PYG{o}{=}\PYG{l+m+mi}{58}\PYG{p}{;}\PYG{n}{DP4}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{;}\PYG{n}{MQ}\PYG{o}{=}\PYG{l+m+mi}{20}\PYG{p}{;}\PYG{n}{FQ}\PYG{o}{=}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{23.4}    \PYG{n}{PL}\PYG{p}{:}\PYG{n}{DP}   \PYG{l+m+mi}{0}\PYG{o}{/}\PYG{l+m+mi}{0}\PYG{p}{:}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{40}\PYG{p}{:}\PYG{l+m+mi}{2}\PYG{p}{:}\PYG{l+m+mi}{4}  \PYG{l+m+mi}{1}\PYG{o}{/}\PYG{l+m+mi}{1}\PYG{p}{:}\PYG{l+m+mi}{38}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{:}\PYG{l+m+mi}{2}\PYG{p}{:}\PYG{l+m+mi}{4}  \PYG{l+m+mi}{1}\PYG{o}{/}\PYG{l+m+mi}{1}\PYG{p}{:}\PYG{l+m+mi}{38}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{:}\PYG{l+m+mi}{2}\PYG{p}{:}\PYG{l+m+mi}{4}  \PYG{l+m+mi}{0}\PYG{o}{/}\PYG{l+m+mi}{0}\PYG{p}{:}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{40}\PYG{p}{:}\PYG{l+m+mi}{2}\PYG{p}{:}\PYG{l+m+mi}{4}
\PYG{n}{ch01}  \PYG{l+m+mi}{108}     \PYG{o}{.}       \PYG{n}{A}       \PYG{n}{C}\PYG{p}{,}\PYG{n}{G}     \PYG{l+m+mi}{999}     \PYG{o}{.}       \PYG{n}{DP}\PYG{o}{=}\PYG{l+m+mi}{52}\PYG{p}{;}\PYG{n}{VDB}\PYG{o}{=}\PYG{l+m+mf}{6.361343e\PYGZhy{}02}\PYG{p}{;}\PYG{n}{RPB}\PYG{o}{=}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{1.264051e+00}\PYG{p}{;}\PYG{n}{AF1}\PYG{o}{=}\PYG{l+m+mf}{0.9325}\PYG{p}{;}\PYG{n}{AC1}\PYG{o}{=}\PYG{l+m+mi}{54}\PYG{p}{;}\PYG{n}{DP4}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{2}\PYG{p}{,}\PYG{l+m+mi}{20}\PYG{p}{,}\PYG{l+m+mi}{26}\PYG{p}{;}\PYG{n}{MQ}\PYG{o}{=}\PYG{l+m+mi}{20}\PYG{p}{;}\PYG{n}{FQ}\PYG{o}{=}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{16.1}\PYG{p}{;}\PYG{n}{PV4}\PYG{o}{=}\PYG{l+m+mf}{0.5}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{,}\PYG{l+m+mi}{1}   \PYG{n}{GT}\PYG{p}{:}\PYG{n}{PL}\PYG{p}{:}\PYG{n}{DP}\PYG{p}{:}\PYG{n}{GQ}     \PYG{l+m+mi}{1}\PYG{o}{/}\PYG{l+m+mi}{1}\PYG{p}{:}\PYG{l+m+mi}{20}\PYG{p}{,}\PYG{l+m+mi}{3}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{20}\PYG{p}{,}\PYG{l+m+mi}{3}\PYG{p}{,}\PYG{l+m+mi}{20}\PYG{p}{:}\PYG{l+m+mi}{1}\PYG{p}{:}\PYG{l+m+mi}{11} \PYG{l+m+mi}{1}\PYG{o}{/}\PYG{l+m+mi}{1}\PYG{p}{:}\PYG{l+m+mi}{36}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{36}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{36}\PYG{p}{:}\PYG{l+m+mi}{2}\PYG{p}{:}\PYG{l+m+mi}{13} \PYG{l+m+mi}{1}\PYG{o}{/}\PYG{l+m+mi}{1}\PYG{p}{:}\PYG{l+m+mi}{36}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{36}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{36}\PYG{p}{:}\PYG{l+m+mi}{2}\PYG{p}{:}\PYG{l+m+mi}{13} \PYG{l+m+mi}{1}\PYG{o}{/}\PYG{l+m+mi}{1}\PYG{p}{:}\PYG{l+m+mi}{95}\PYG{p}{,}\PYG{l+m+mi}{95}\PYG{p}{,}\PYG{l+m+mi}{95}\PYG{p}{,}\PYG{l+m+mi}{18}\PYG{p}{,}\PYG{l+m+mi}{18}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{:}\PYG{l+m+mi}{6}\PYG{p}{:}\PYG{l+m+mi}{8}
\end{sphinxVerbatim}


\chapter{Program Parameter Descriptions}
\label{\detokenize{prog_desc::doc}}\label{\detokenize{prog_desc:program-parameter-descriptions}}

\section{fasta2mvf}
\label{\detokenize{prog_desc:fasta2mvf}}

\subsection{Description}
\label{\detokenize{prog_desc:description}}
This program is used to convert a FASTA file into MVF format.


\subsection{Parameters}
\label{\detokenize{prog_desc:parameters}}

\subsubsection{\sphinxstyleliteralintitle{-h/-{-}help}}
\label{\detokenize{prog_desc:h-help}}
\sphinxstylestrong{Description:} show this help message and exit

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-i/-{-}fasta} (required)}
\label{\detokenize{prog_desc:i-fasta-required}}
\sphinxstylestrong{Description:} input FASTA file(s)

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-o/-{-}out} (required)}
\label{\detokenize{prog_desc:o-out-required}}
\sphinxstylestrong{Description:} output MVF file

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-c/-{-}contig-ids/-{-}contigids}}
\label{\detokenize{prog_desc:c-contig-ids-contigids}}
\sphinxstylestrong{Description:} manually specify one or more contig ids as ID:NAME

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}contig-by-file/-{-}contigbyfile}}
\label{\detokenize{prog_desc:contig-by-file-contigbyfile}}
\sphinxstylestrong{Description:} Contigs are designated by separate files.

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}contig-field/-{-}contigfield}}
\label{\detokenize{prog_desc:contig-field-contigfield}}
\sphinxstylestrong{Description:} When headers are split by \textendash{}field-sep, the 0-based index of the contig id.

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-f/-{-}flavor}}
\label{\detokenize{prog_desc:f-flavor}}
\sphinxstylestrong{Description:} type of file {[}dna{]} or protein

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} dna

\sphinxstylestrong{Choices:} {[}‘dna’, ‘protein’{]}


\subsubsection{\sphinxstyleliteralintitle{-{-}manual-coord/-{-}manualcoord}}
\label{\detokenize{prog_desc:manual-coord-manualcoord}}
\sphinxstylestrong{Description:} manually specify reference coordinates for each file in the format CONTIGID:START..STOP, …

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}overwrite}}
\label{\detokenize{prog_desc:overwrite}}
\sphinxstylestrong{Description:} USE WITH CAUTION: force overwrite of outputs

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-s/-{-}sample-replace/-{-}samplereplace}}
\label{\detokenize{prog_desc:s-sample-replace-samplereplace}}
\sphinxstylestrong{Description:} one or more \sphinxurl{TAG:NEWLABEL} or TAG, items, if TAG found in sample label, replace with NEW (or TAG if NEW not specified) NEW and TAG must each be unique

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}sample-field/-{-}samplefield}}
\label{\detokenize{prog_desc:sample-field-samplefield}}
\sphinxstylestrong{Description:} when headers are split by \textendash{}field-sep, the 0-based index of the sample id

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-B/-{-}read-buffer/-{-}readbuffer}}
\label{\detokenize{prog_desc:b-read-buffer-readbuffer}}
\sphinxstylestrong{Description:} number of lines to hold in READ buffer

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 100000


\subsubsection{\sphinxstyleliteralintitle{-F/-{-}field-sep/-{-}fieldsep}}
\label{\detokenize{prog_desc:f-field-sep-fieldsep}}
\sphinxstylestrong{Description:} FASTA field separator; assumes ‘\textgreater{}database accession locus’ format

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None

\sphinxstylestrong{Choices:} {[}‘TAB’, ‘SPACE’, ‘DBLSPACE’, ‘COMMA’, ‘MIXED’, ‘PIPE’, ‘AT’, ‘UNDER’, ‘DBLUNDER’{]}


\subsubsection{\sphinxstyleliteralintitle{-R/-{-}ref-label/-{-}reflabel}}
\label{\detokenize{prog_desc:r-ref-label-reflabel}}
\sphinxstylestrong{Description:} label for reference sample

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} REF


\subsubsection{\sphinxstyleliteralintitle{-W/-{-}write-buffer/-{-}writebuffer}}
\label{\detokenize{prog_desc:w-write-buffer-writebuffer}}
\sphinxstylestrong{Description:} number of lines to hold in WRITE buffer

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 100000


\section{maf2mvf}
\label{\detokenize{prog_desc:maf2mvf}}

\subsection{Description}
\label{\detokenize{prog_desc:id1}}
This program analyzes a DNA MVF alignment using the modules specified below,
use the \textendash{}morehelp option for additional module information.


\subsection{Parameters}
\label{\detokenize{prog_desc:id2}}

\subsubsection{\sphinxstyleliteralintitle{-h/-{-}help}}
\label{\detokenize{prog_desc:id3}}
\sphinxstylestrong{Description:} show this help message and exit

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-i/-{-}maf} (required)}
\label{\detokenize{prog_desc:i-maf-required}}
\sphinxstylestrong{Description:} input MAF file

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-o/-{-}out} (required)}
\label{\detokenize{prog_desc:id4}}
\sphinxstylestrong{Description:} output MVF file

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}overwrite}}
\label{\detokenize{prog_desc:id5}}
\sphinxstylestrong{Description:} None

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-s/-{-}sample-tags/-{-}sampletags}}
\label{\detokenize{prog_desc:s-sample-tags-sampletags}}
\sphinxstylestrong{Description:} one or more \sphinxurl{TAG:NEWLABEL} or TAG, items, if TAG found in sample label, replace with NEW (or TAG if NEW not specified) NEW and TAG must each be unique.

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-B/-{-}line-buffer/-{-}linebuffer}}
\label{\detokenize{prog_desc:b-line-buffer-linebuffer}}
\sphinxstylestrong{Description:} number of lines to hold in read/write buffer

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 100000


\subsubsection{\sphinxstyleliteralintitle{-M/-{-}mvf-ref-label/-{-}mvfreflabel}}
\label{\detokenize{prog_desc:m-mvf-ref-label-mvfreflabel}}
\sphinxstylestrong{Description:} new label for reference sample (default=’REF’)

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} REF


\subsubsection{\sphinxstyleliteralintitle{-R/-{-}ref-tag/-{-}reftag}}
\label{\detokenize{prog_desc:r-ref-tag-reftag}}
\sphinxstylestrong{Description:} old reference tag

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\section{mvf2dump}
\label{\detokenize{prog_desc:mvf2dump}}

\subsection{Description}
\label{\detokenize{prog_desc:id6}}
This program exports the entirety of an MVF to FASTA format,
with many fewer options than mvf2fasta.py.  This is designed
to export large MVF files faster, but with less specific
formatting and region-finding options.


\subsection{Parameters}
\label{\detokenize{prog_desc:id7}}

\subsubsection{\sphinxstyleliteralintitle{-h/-{-}help}}
\label{\detokenize{prog_desc:id8}}
\sphinxstylestrong{Description:} show this help message and exit

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-i/-{-}mvf} (required)}
\label{\detokenize{prog_desc:i-mvf-required}}
\sphinxstylestrong{Description:} Input MVF file.

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-o/-{-}outprefix} (required)}
\label{\detokenize{prog_desc:o-outprefix-required}}
\sphinxstylestrong{Description:} Target FASTA file

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-d/-{-}outdata}}
\label{\detokenize{prog_desc:d-outdata}}
\sphinxstylestrong{Description:} output dna, rna or prot data

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None

\sphinxstylestrong{Choices:} (‘dna’, ‘rna’, ‘prot’)


\subsubsection{\sphinxstyleliteralintitle{-{-}quiet}}
\label{\detokenize{prog_desc:quiet}}
\sphinxstylestrong{Description:} suppress screen output

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-s/-{-}samples}}
\label{\detokenize{prog_desc:s-samples}}
\sphinxstylestrong{Description:} One or more taxon labels, leave blank for all

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-t/-{-}tmpdir}}
\label{\detokenize{prog_desc:t-tmpdir}}
\sphinxstylestrong{Description:} directory to write temporary fasta files

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} .


\subsubsection{\sphinxstyleliteralintitle{-B/-{-}buffer}}
\label{\detokenize{prog_desc:b-buffer}}
\sphinxstylestrong{Description:} size (Mbp) of write buffer for each sample

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 10


\section{mvf2fasta}
\label{\detokenize{prog_desc:mvf2fasta}}

\subsection{Description}
\label{\detokenize{prog_desc:id9}}
This program takes an MVF file and converts the data to a FASTA file


\subsection{Parameters}
\label{\detokenize{prog_desc:id10}}

\subsubsection{\sphinxstyleliteralintitle{-h/-{-}help}}
\label{\detokenize{prog_desc:id11}}
\sphinxstylestrong{Description:} show this help message and exit

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-i/-{-}mvf} (required)}
\label{\detokenize{prog_desc:id12}}
\sphinxstylestrong{Description:} Input MVF file.

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-o/-{-}out} (required)}
\label{\detokenize{prog_desc:id13}}
\sphinxstylestrong{Description:} target FASTA file

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-r/-{-}regions} (required)}
\label{\detokenize{prog_desc:r-regions-required}}
\sphinxstylestrong{Description:} One or more space-separated arguments formatted as contigid,start,stop coordinates are (inclusive)

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-d/-{-}outdata}}
\label{\detokenize{prog_desc:id14}}
\sphinxstylestrong{Description:} Output dna, rna or prot data.

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None

\sphinxstylestrong{Choices:} (‘dna’, ‘rna’, ‘prot’)


\subsubsection{\sphinxstyleliteralintitle{-l/-{-}labeltype}}
\label{\detokenize{prog_desc:l-labeltype}}
\sphinxstylestrong{Description:} Long labels with all metadata or short ids

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} long

\sphinxstylestrong{Choices:} (‘long’, ‘short’)


\subsubsection{\sphinxstyleliteralintitle{-{-}quiet}}
\label{\detokenize{prog_desc:id15}}
\sphinxstylestrong{Description:} suppress screen output

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-s/-{-}samples}}
\label{\detokenize{prog_desc:id16}}
\sphinxstylestrong{Description:} One or more taxon labels, leave blank for all

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-t/-{-}tmpdir}}
\label{\detokenize{prog_desc:id17}}
\sphinxstylestrong{Description:} directory to write temporary fasta files

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} .


\subsubsection{\sphinxstyleliteralintitle{-B/-{-}buffer}}
\label{\detokenize{prog_desc:id18}}
\sphinxstylestrong{Description:} size (Mbp) of write buffer for each sample

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 10


\section{mvf2phy}
\label{\detokenize{prog_desc:mvf2phy}}

\subsection{Description}
\label{\detokenize{prog_desc:id19}}
This program is used to export MVF data to Phylip format.


\subsection{Parameters}
\label{\detokenize{prog_desc:id20}}

\subsubsection{\sphinxstyleliteralintitle{-h/-{-}help}}
\label{\detokenize{prog_desc:id21}}
\sphinxstylestrong{Description:} show this help message and exit

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-i/-{-}mvf} (required)}
\label{\detokenize{prog_desc:id22}}
\sphinxstylestrong{Description:} Input MVF file.

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-o/-{-}out} (required)}
\label{\detokenize{prog_desc:id23}}
\sphinxstylestrong{Description:} Output Phylip file.

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-d/-{-}outdata}}
\label{\detokenize{prog_desc:id24}}
\sphinxstylestrong{Description:} Output dna, rna or prot data.

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None

\sphinxstylestrong{Choices:} (‘dna’, ‘rna’, ‘prot’)


\subsubsection{\sphinxstyleliteralintitle{-p/-{-}partition}}
\label{\detokenize{prog_desc:p-partition}}
\sphinxstylestrong{Description:} Output a CSV partitions file with RAxMLformatting for use in partitioned phylogenetic methods.

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}quiet}}
\label{\detokenize{prog_desc:id25}}
\sphinxstylestrong{Description:} suppress screen output

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-r/-{-}region}}
\label{\detokenize{prog_desc:r-region}}
\sphinxstylestrong{Description:} Path of a plain text file containing one more lines with entries ‘contigid,stop,start’ (one per line, inclusive coordinates) all data will be returned if left blank.

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-s/-{-}samples}}
\label{\detokenize{prog_desc:id26}}
\sphinxstylestrong{Description:} One or more taxon labels, leave blank for all.

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-t/-{-}tmpdir}}
\label{\detokenize{prog_desc:id27}}
\sphinxstylestrong{Description:} directory to write temporary fasta files

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} .


\subsubsection{\sphinxstyleliteralintitle{-B/-{-}buffer}}
\label{\detokenize{prog_desc:id28}}
\sphinxstylestrong{Description:} size (bp) of write buffer for each sample

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 100000


\subsubsection{\sphinxstyleliteralintitle{-L/-{-}labeltype}}
\label{\detokenize{prog_desc:id29}}
\sphinxstylestrong{Description:} Long labels with all metadata or short ids

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} short

\sphinxstylestrong{Choices:} (‘long’, ‘short’)


\section{mvf\_analyze\_codon}
\label{\detokenize{prog_desc:mvf-analyze-codon}}

\subsection{Description}
\label{\detokenize{prog_desc:id30}}
This program analyzes a codon MVF using several analysis modules.
Run ‘python3 mvf\_analyze\_codon.py \textendash{}morehelp’ for details on
module functions.


\subsection{Parameters}
\label{\detokenize{prog_desc:id31}}

\subsubsection{module}
\label{\detokenize{prog_desc:module}}
\sphinxstylestrong{Description:} None

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None

\sphinxstylestrong{Choices:} (‘Coverage’, ‘GroupUniqueAlleleWindow’, ‘PiDiversityWindow’, ‘PairwiseNS’)


\subsubsection{\sphinxstyleliteralintitle{-h/-{-}help}}
\label{\detokenize{prog_desc:id32}}
\sphinxstylestrong{Description:} show this help message and exit

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}all-sample-trees/-{-}allsampletrees}}
\label{\detokenize{prog_desc:all-sample-trees-allsampletrees}}
\sphinxstylestrong{Description:} (GroupUniqueAlleleWindow) Makes trees from all samples instead of only the most complete sequence from each species

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}allele-groups/-{-}allelegroups}}
\label{\detokenize{prog_desc:allele-groups-allelegroups}}\begin{description}
\item[{\sphinxstylestrong{Description:} GROUP1:LABEL,LABEL GROUP2:LABEL,LABEL}] \leavevmode
(GroupUniqueAlleleWindow)

\end{description}

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}branchlrt}}
\label{\detokenize{prog_desc:branchlrt}}
\sphinxstylestrong{Description:} (GroupUniqueAlleleWindow) Specify the output file for and turn on the RAxML-PAML format LRT test scan for selection on the target branch in addition to the basic patterns scan

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-c/-{-}contigs}}
\label{\detokenize{prog_desc:c-contigs}}
\sphinxstylestrong{Description:} List of space-separated contig ids.

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-g/-{-}gff}}
\label{\detokenize{prog_desc:g-gff}}
\sphinxstylestrong{Description:} GFF3 file for use in annotation

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-i/-{-}mvf}}
\label{\detokenize{prog_desc:i-mvf}}
\sphinxstylestrong{Description:} Input MVF file.

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-m/-{-}mincoverage}}
\label{\detokenize{prog_desc:m-mincoverage}}
\sphinxstylestrong{Description:} Minimum number of samples with alleles needed to use site for analysis.

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}morehelp}}
\label{\detokenize{prog_desc:morehelp}}
\sphinxstylestrong{Description:} Get additional information on modules.

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}num-target-species/-{-}targetspec}}
\label{\detokenize{prog_desc:num-target-species-targetspec}}
\sphinxstylestrong{Description:} (GroupUniqueAlleleWindow) Specify the minimum number of taxa in the target set that are required to conduct analysis

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 1


\subsubsection{\sphinxstyleliteralintitle{-o/-{-}out}}
\label{\detokenize{prog_desc:o-out}}
\sphinxstylestrong{Description:} output file

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}output-align/-{-}outputalign}}
\label{\detokenize{prog_desc:output-align-outputalign}}
\sphinxstylestrong{Description:} (GroupUniqueAlleleWindow) Output alignment to this file path in phylip format.

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}pamltmp}}
\label{\detokenize{prog_desc:pamltmp}}
\sphinxstylestrong{Description:} path for temporary folder for PAML output files

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} pamltmp


\subsubsection{\sphinxstyleliteralintitle{-s/-{-}samples}}
\label{\detokenize{prog_desc:id33}}
\sphinxstylestrong{Description:} List of space-separated sample names.

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}species-groups/-{-}speciesgroups}}
\label{\detokenize{prog_desc:species-groups-speciesgroups}}
\sphinxstylestrong{Description:} None

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}target}}
\label{\detokenize{prog_desc:target}}
\sphinxstylestrong{Description:} (GroupUniqueAlleleWindow) Specify the taxa labels that define the target lineage-specific branch to be tested.

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-w/-{-}windowsize}}
\label{\detokenize{prog_desc:w-windowsize}}
\sphinxstylestrong{Description:} Window size in bp, use -1 for whole contig.

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} -1


\subsubsection{\sphinxstyleliteralintitle{-x/-{-}chi-test/-{-}chitest}}
\label{\detokenize{prog_desc:x-chi-test-chitest}}
\sphinxstylestrong{Description:} (GroupUniqueAlleleWindow,PairwiseDNDS)Input two number values for expected Nonsynonymous and Synonymous expected values.

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-E/-{-}endcontig}}
\label{\detokenize{prog_desc:e-endcontig}}
\sphinxstylestrong{Description:} Numerical id for the ending contig.

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 100000000


\subsubsection{\sphinxstyleliteralintitle{-L/-{-}uselabels}}
\label{\detokenize{prog_desc:l-uselabels}}
\sphinxstylestrong{Description:} Use contig labels instead of IDs in output.

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-O/-{-}outgroup}}
\label{\detokenize{prog_desc:o-outgroup}}
\sphinxstylestrong{Description:} (GroupUniqueAlleleWindow) Specify sample name with which to root trees.

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-P/-{-}codemlpath}}
\label{\detokenize{prog_desc:p-codemlpath}}
\sphinxstylestrong{Description:} Full path for PAML codeml executable.

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} codeml


\subsubsection{\sphinxstyleliteralintitle{-S/-{-}startcontig}}
\label{\detokenize{prog_desc:s-startcontig}}
\sphinxstylestrong{Description:} Numerical ID for the starting contig.

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 0


\subsubsection{\sphinxstyleliteralintitle{-X/-{-}raxmlpath}}
\label{\detokenize{prog_desc:x-raxmlpath}}
\sphinxstylestrong{Description:} Full path to RAxML program executable.

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} raxml


\section{mvf\_analyze\_dna}
\label{\detokenize{prog_desc:mvf-analyze-dna}}

\subsection{Description}
\label{\detokenize{prog_desc:id34}}
This program analyzes a DNA MVF alignment using the modules specified below,
use the \textendash{}morehelp option for additional module information.


\subsection{Parameters}
\label{\detokenize{prog_desc:id35}}

\subsubsection{module}
\label{\detokenize{prog_desc:id36}}
\sphinxstylestrong{Description:} analysis module to run

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None

\sphinxstylestrong{Choices:} (‘BaseCountWindow’, ‘Coverage’, ‘DstatComb’, ‘PairwiseDistance’, ‘PairwiseDistanceWindow’, ‘PatternCount’)


\subsubsection{\sphinxstyleliteralintitle{-h/-{-}help}}
\label{\detokenize{prog_desc:id37}}
\sphinxstylestrong{Description:} show this help message and exit

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-i/-{-}mvf} (required)}
\label{\detokenize{prog_desc:id38}}
\sphinxstylestrong{Description:} Input MVF file.

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-o/-{-}out} (required)}
\label{\detokenize{prog_desc:id39}}
\sphinxstylestrong{Description:} output file

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}base-match}}
\label{\detokenize{prog_desc:base-match}}
\sphinxstylestrong{Description:} {[}BaseCountWindow{]} string of bases to match (i.e. numerator).

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}base-total}}
\label{\detokenize{prog_desc:base-total}}
\sphinxstylestrong{Description:} {[}BaseCountWindow{]} string of bases for total (i.e. denominator).

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-c/-{-}contigs}}
\label{\detokenize{prog_desc:id40}}
\sphinxstylestrong{Description:} limit analyses to these contigs

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-m/-{-}mincoverage}}
\label{\detokenize{prog_desc:id41}}
\sphinxstylestrong{Description:} mininum sample coverage for site

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}morehelp}}
\label{\detokenize{prog_desc:id42}}
\sphinxstylestrong{Description:} get additional information on modules

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-s/-{-}samples}}
\label{\detokenize{prog_desc:id43}}
\sphinxstylestrong{Description:} limit analyses to these samples

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-w/-{-}windowsize}}
\label{\detokenize{prog_desc:id44}}
\sphinxstylestrong{Description:} window size

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 100000


\section{mvf\_annotate}
\label{\detokenize{prog_desc:mvf-annotate}}

\subsection{Description}
\label{\detokenize{prog_desc:id45}}
This program takes a DNA MVF alignment and annotates the output into
gene boudaries.


\subsection{Parameters}
\label{\detokenize{prog_desc:id46}}

\subsubsection{\sphinxstyleliteralintitle{-h/-{-}help}}
\label{\detokenize{prog_desc:id47}}
\sphinxstylestrong{Description:} show this help message and exit

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-g/-{-}gff}}
\label{\detokenize{prog_desc:id48}}
\sphinxstylestrong{Description:} Input gff annotation file.

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-i/-{-}mvf}}
\label{\detokenize{prog_desc:id49}}
\sphinxstylestrong{Description:} Input MVF file.

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-o/-{-}out}}
\label{\detokenize{prog_desc:id50}}
\sphinxstylestrong{Description:} Output annotated MVF file

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}overwrite}}
\label{\detokenize{prog_desc:id51}}
\sphinxstylestrong{Description:} USE WITH CAUTION: force overwrite of outputs

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}quiet}}
\label{\detokenize{prog_desc:id52}}
\sphinxstylestrong{Description:} suppress progress meter

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-B/-{-}linebuffer}}
\label{\detokenize{prog_desc:b-linebuffer}}
\sphinxstylestrong{Description:} Number of entries to store in memory at a time.

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 100000


\subsubsection{\sphinxstyleliteralintitle{-F/-{-}filter\_annotation}}
\label{\detokenize{prog_desc:f-filter-annotation}}
\sphinxstylestrong{Description:} Skip entries in the GFF file that contain this string in their ‘Notes’

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-M/-{-}nongenic-margin}}
\label{\detokenize{prog_desc:m-nongenic-margin}}
\sphinxstylestrong{Description:} for \textendash{}unnanotated-mode, only retain positions that are this number of bp away from an annotated region boundary

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 0


\subsubsection{\sphinxstyleliteralintitle{-N/-{-}nongenic-mode}}
\label{\detokenize{prog_desc:n-nongenic-mode}}
\sphinxstylestrong{Description:} Instead of returning annotated genes, return the non-genic regions without without changing contigs or coordinates

\sphinxstylestrong{Type:} boolean flag


\section{mvf\_check}
\label{\detokenize{prog_desc:mvf-check}}

\subsection{Description}
\label{\detokenize{prog_desc:id53}}
This program checks an MVF file for inconsistencies or errors


\subsection{Parameters}
\label{\detokenize{prog_desc:id54}}

\subsubsection{mvf}
\label{\detokenize{prog_desc:mvf}}
\sphinxstylestrong{Description:} Input MVF file.

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-h/-{-}help}}
\label{\detokenize{prog_desc:id55}}
\sphinxstylestrong{Description:} show this help message and exit

\sphinxstylestrong{Type:} boolean flag


\section{mvf\_chromoplot}
\label{\detokenize{prog_desc:mvf-chromoplot}}

\subsection{Description}
\label{\detokenize{prog_desc:id56}}
This program creates a chromoplot from an MVF alignment.
A chromoplot shows a genome-wide diagram of different
evolutionary histories for a given quartet of taxa.


\subsection{Parameters}
\label{\detokenize{prog_desc:id57}}

\subsubsection{\sphinxstyleliteralintitle{-h/-{-}help}}
\label{\detokenize{prog_desc:id58}}
\sphinxstylestrong{Description:} show this help message and exit

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-i/-{-}mvf} (required)}
\label{\detokenize{prog_desc:id59}}
\sphinxstylestrong{Description:} Input MVF file.

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-s/-{-}samples} (required)}
\label{\detokenize{prog_desc:s-samples-required}}
\sphinxstylestrong{Description:} 3 or more taxa to use for quartets

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-G/-{-}outgroup} (required)}
\label{\detokenize{prog_desc:g-outgroup-required}}
\sphinxstylestrong{Description:} 1 or more outgroups to use for quartets

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-c/-{-}contigs}}
\label{\detokenize{prog_desc:id60}}
\sphinxstylestrong{Description:} Enter the ids of one or more contigs in the order they will appear in the chromoplot. (defaults to all ids in order present in MVF)

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-o/-{-}outprefix}}
\label{\detokenize{prog_desc:o-outprefix}}
\sphinxstylestrong{Description:} Output prefix (not required).

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-q/-{-}quiet}}
\label{\detokenize{prog_desc:q-quiet}}
\sphinxstylestrong{Description:} suppress all output messages

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-w/-{-}windowsize}}
\label{\detokenize{prog_desc:id61}}
\sphinxstylestrong{Description:} None

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 100000


\subsubsection{\sphinxstyleliteralintitle{-x/-{-}xscale}}
\label{\detokenize{prog_desc:x-xscale}}
\sphinxstylestrong{Description:} Width (in number of pixels) for each window

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 1


\subsubsection{\sphinxstyleliteralintitle{-y/-{-}yscale}}
\label{\detokenize{prog_desc:y-yscale}}
\sphinxstylestrong{Description:} Height (in number of pixels) for each track

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 20


\subsubsection{\sphinxstyleliteralintitle{-C/-{-}colors}}
\label{\detokenize{prog_desc:c-colors}}
\sphinxstylestrong{Description:} three colors to use for chromoplot

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None

\sphinxstylestrong{Choices:} \{‘lgrey’: (250, 250, 250), ‘dgrey’: (192, 192, 192), ‘black’: (0, 0, 0), ‘white’: (255, 255, 255), ‘red’: (192, 0, 0), ‘orange’: (217, 95, 2), ‘yellow’: (192, 192, 0), ‘green’: (0, 192, 0), ‘blue’: (0, 0, 192), ‘teal’: (27, 158, 119), ‘puce’: (117, 112, 179), ‘purple’: (192, 0, 192), ‘none’: ()\}


\subsubsection{\sphinxstyleliteralintitle{-E/-{-}emptymask}}
\label{\detokenize{prog_desc:e-emptymask}}
\sphinxstylestrong{Description:} Mask empty regions with this color.

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} none

\sphinxstylestrong{Choices:} \{‘lgrey’: (250, 250, 250), ‘dgrey’: (192, 192, 192), ‘black’: (0, 0, 0), ‘white’: (255, 255, 255), ‘red’: (192, 0, 0), ‘orange’: (217, 95, 2), ‘yellow’: (192, 192, 0), ‘green’: (0, 192, 0), ‘blue’: (0, 0, 192), ‘teal’: (27, 158, 119), ‘puce’: (117, 112, 179), ‘purple’: (192, 0, 192), ‘none’: ()\}


\subsubsection{\sphinxstyleliteralintitle{-I/-{-}infotrack}}
\label{\detokenize{prog_desc:i-infotrack}}
\sphinxstylestrong{Description:} Include an additional coverage information track that will show empty, uninformative, and informative loci. (Useful for ranscriptomes/RAD or other reduced sampling.

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-M/-{-}majority}}
\label{\detokenize{prog_desc:m-majority}}
\sphinxstylestrong{Description:} Plot only 100\% shading in the majority track  rather than shaded proportions in all tracks.

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-P/-{-}plottype}}
\label{\detokenize{prog_desc:p-plottype}}
\sphinxstylestrong{Description:} PNG image (default) or graph via matplotlib (experimental)

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} image

\sphinxstylestrong{Choices:} {[}‘graph’, ‘image’{]}


\section{mvf\_filter}
\label{\detokenize{prog_desc:mvf-filter}}

\subsection{Description}
\label{\detokenize{prog_desc:id62}}
This program filters an MVF alignment using the modules specified below,
use the \textendash{}morehelp option for additional module information.


\subsection{Parameters}
\label{\detokenize{prog_desc:id63}}

\subsubsection{\sphinxstyleliteralintitle{-h/-{-}help}}
\label{\detokenize{prog_desc:id64}}
\sphinxstylestrong{Description:} show this help message and exit

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-a/-{-}actions}}
\label{\detokenize{prog_desc:a-actions}}
\sphinxstylestrong{Description:} set of actions:args to perform, note these are done in order as listed

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-i/-{-}mvf}}
\label{\detokenize{prog_desc:id65}}
\sphinxstylestrong{Description:} Input MVF file.

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-l/-{-}labels}}
\label{\detokenize{prog_desc:l-labels}}
\sphinxstylestrong{Description:} use sample labels instead of indices

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}morehelp}}
\label{\detokenize{prog_desc:id66}}
\sphinxstylestrong{Description:} prints full module list and descriptions

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-o/-{-}out}}
\label{\detokenize{prog_desc:id67}}
\sphinxstylestrong{Description:} Output MVF file

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}overwrite}}
\label{\detokenize{prog_desc:id68}}
\sphinxstylestrong{Description:} USE WITH CAUTION: force overwrite of outputs

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-q/-{-}quiet}}
\label{\detokenize{prog_desc:id69}}
\sphinxstylestrong{Description:} suppress progress meter

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}test}}
\label{\detokenize{prog_desc:test}}
\sphinxstylestrong{Description:} manually input a line for testing

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}test-nchar}}
\label{\detokenize{prog_desc:test-nchar}}
\sphinxstylestrong{Description:} total number of samples for test string

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-B/-{-}linebuffer}}
\label{\detokenize{prog_desc:id70}}
\sphinxstylestrong{Description:} number of lines to write at once to MVF

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 100000


\subsubsection{\sphinxstyleliteralintitle{-V/-{-}verbose}}
\label{\detokenize{prog_desc:v-verbose}}
\sphinxstylestrong{Description:} report every line (for debugging)

\sphinxstylestrong{Type:} boolean flag


\section{mvf\_join}
\label{\detokenize{prog_desc:mvf-join}}

\subsection{Description}
\label{\detokenize{prog_desc:id71}}
This program checks an MVF file for inconsistencies or errors


\subsection{Parameters}
\label{\detokenize{prog_desc:id72}}

\subsubsection{\sphinxstyleliteralintitle{-h/-{-}help}}
\label{\detokenize{prog_desc:id73}}
\sphinxstylestrong{Description:} show this help message and exit

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-i/-{-}mvf} (required)}
\label{\detokenize{prog_desc:id74}}
\sphinxstylestrong{Description:} One or more mvf files.

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-o-{-}out} (required)}
\label{\detokenize{prog_desc:id75}}
\sphinxstylestrong{Description:} Output mvf file.

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-c/-{-}newcontigs}}
\label{\detokenize{prog_desc:c-newcontigs}}
\sphinxstylestrong{Description:} By default, contigs are matched between files using their text labels in the header. Use this option to turn matching off and treat each file’s contigs as distinct.

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}overwrite}}
\label{\detokenize{prog_desc:id76}}
\sphinxstylestrong{Description:} USE WITH CAUTION: force overwrite of outputs

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}quiet}}
\label{\detokenize{prog_desc:id77}}
\sphinxstylestrong{Description:} suppress progress meter

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-s/-{-}newsamples}}
\label{\detokenize{prog_desc:s-newsamples}}
\sphinxstylestrong{Description:} By default, samples are matched between files using their text labels in the header. Use this option to turn matching off and treat each file’s sample columns as distinct.

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-B/-{-}linebuffer}}
\label{\detokenize{prog_desc:id78}}
\sphinxstylestrong{Description:} number of entries to write in a block

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 100000


\subsubsection{\sphinxstyleliteralintitle{-M/-{-}main\_header\_file}}
\label{\detokenize{prog_desc:m-main-header-file}}
\sphinxstylestrong{Description:} Output file will use same headers as this input file (default=first in list).

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\section{mvf\_translate}
\label{\detokenize{prog_desc:mvf-translate}}

\subsection{Description}
\label{\detokenize{prog_desc:id79}}
This program translates a DNA MVF file into a codon or protein MVF file
using a GFF3 annotation file.


\subsection{Parameters}
\label{\detokenize{prog_desc:id80}}

\subsubsection{\sphinxstyleliteralintitle{-h/-{-}help}}
\label{\detokenize{prog_desc:id81}}
\sphinxstylestrong{Description:} show this help message and exit

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-i/-{-}mvf} (required)}
\label{\detokenize{prog_desc:id82}}
\sphinxstylestrong{Description:} Input MAF file

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-o/-{-}out} (required)}
\label{\detokenize{prog_desc:id83}}
\sphinxstylestrong{Description:} Output MVF file

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-g/-{-}gff}}
\label{\detokenize{prog_desc:id84}}
\sphinxstylestrong{Description:} Input GFF3 file. If GFF3 not provided, alignments are assumed to be in-frame coding sequences.

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}overwrite}}
\label{\detokenize{prog_desc:id85}}
\sphinxstylestrong{Description:} USE WITH CAUTION: force overwrite of outputs

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}quiet}}
\label{\detokenize{prog_desc:id86}}
\sphinxstylestrong{Description:} suppress progress meter

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-t-{-}outtype}}
\label{\detokenize{prog_desc:t-outtype}}
\sphinxstylestrong{Description:} protein=single data column of protein alleles; codon=four columns with: protein frame1 frame2 frame3

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} codon

\sphinxstylestrong{Choices:} {[}‘protein’, ‘codon’{]}


\subsubsection{\sphinxstyleliteralintitle{-B/-{-}line-buffer/-{-}linebuffer}}
\label{\detokenize{prog_desc:id87}}
\sphinxstylestrong{Description:} number of entries to write in a block

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 100000


\subsubsection{\sphinxstyleliteralintitle{-F/-{-}filter-annotation}}
\label{\detokenize{prog_desc:id88}}
\sphinxstylestrong{Description:} skip GFF entries with text matching this in their ‘Notes’ field

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\section{mvf\_window\_tree}
\label{\detokenize{prog_desc:mvf-window-tree}}

\subsection{Description}
\label{\detokenize{prog_desc:id89}}
This program makes phylogenies from individual genomic windows of
a DNA MVF alignment (Requires: BioPython).


\subsection{Parameters}
\label{\detokenize{prog_desc:id90}}

\subsubsection{\sphinxstyleliteralintitle{-h/-{-}help}}
\label{\detokenize{prog_desc:id91}}
\sphinxstylestrong{Description:} show this help message and exit

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-i/-{-}mvf} (required)}
\label{\detokenize{prog_desc:id92}}
\sphinxstylestrong{Description:} Input MVF file.

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-o/-{-}out} (required)}
\label{\detokenize{prog_desc:id93}}
\sphinxstylestrong{Description:} Tree list output text file.

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-b/-{-}bootstrap}}
\label{\detokenize{prog_desc:b-bootstrap}}
\sphinxstylestrong{Description:} turn on rapid bootstrapping for RAxML and perform specified number of replicates

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-c/-{-}contigs}}
\label{\detokenize{prog_desc:id94}}
\sphinxstylestrong{Description:} Contig ids to use in analysis (default=all)

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-d/-{-}duplicateseq}}
\label{\detokenize{prog_desc:d-duplicateseq}}
\sphinxstylestrong{Description:} dontuse=remove duplicate sequences prior to RAxML tree inference, then add them to the tree manually as zero-branch-length sister taxa; keep=keep in for RAxML tree inference (may cause errors for RAxML); remove=remove entirely from alignment

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} dontuse

\sphinxstylestrong{Choices:} {[}‘dontuse’, ‘keep’, ‘remove’{]}


\subsubsection{\sphinxstyleliteralintitle{-e/-{-}outputempty}}
\label{\detokenize{prog_desc:e-outputempty}}
\sphinxstylestrong{Description:} Include entries of windows with no data in output.

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-g/-{-}raxml-outgroups/-{-}raxml\_outgroups}}
\label{\detokenize{prog_desc:g-raxml-outgroups-raxml-outgroups}}
\sphinxstylestrong{Description:} Outgroups taxon labels to use in RAxML.

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-m/-{-}raxml\_model}}
\label{\detokenize{prog_desc:m-raxml-model}}
\sphinxstylestrong{Description:} choose RAxML model

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} GTRGAMMA


\subsubsection{\sphinxstyleliteralintitle{-{-}outputcontiglabels}}
\label{\detokenize{prog_desc:outputcontiglabels}}
\sphinxstylestrong{Description:} Output will use contig labels instead of id numbers.

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}quiet}}
\label{\detokenize{prog_desc:id95}}
\sphinxstylestrong{Description:} suppress screen output

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-r/-{-}rootwith}}
\label{\detokenize{prog_desc:r-rootwith}}
\sphinxstylestrong{Description:} Root output trees with these taxa after RAxML.

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-s/-{-}samples}}
\label{\detokenize{prog_desc:id96}}
\sphinxstylestrong{Description:} One or more taxon labels (default=all)

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}tempdir}}
\label{\detokenize{prog_desc:tempdir}}
\sphinxstylestrong{Description:} Temporary directory path

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} ./raxmltemp


\subsubsection{\sphinxstyleliteralintitle{-{-}tempprefix}}
\label{\detokenize{prog_desc:tempprefix}}
\sphinxstylestrong{Description:} Temporary file prefix

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} mvftree


\subsubsection{\sphinxstyleliteralintitle{-w/-{-}windowsize}}
\label{\detokenize{prog_desc:id97}}
\sphinxstylestrong{Description:} specify genomic region size, or use -1 for whole contig

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 10000


\subsubsection{\sphinxstyleliteralintitle{-A/-{-}choose\_allele/-{-}hapmode}}
\label{\detokenize{prog_desc:a-choose-allele-hapmode}}
\sphinxstylestrong{Description:} Chooses how heterozygous alleles are handled. (none=no splitting (default); randomone=pick one allele randomly (recommended); randomboth=pick two alleles randomly, but keep both; major=pick the more common allele; minor=pick the less common allele; majorminor= pick the major in ‘a’ and minor in ‘b’

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} none

\sphinxstylestrong{Choices:} {[}‘none’, ‘randomone’, ‘randomboth’, ‘major’, ‘minor’, ‘majorminor’{]}


\subsubsection{\sphinxstyleliteralintitle{-C/-{-}minseqcoverage}}
\label{\detokenize{prog_desc:c-minseqcoverage}}\begin{description}
\item[{\sphinxstylestrong{Description:} proportion of total alignment a sequence}] \leavevmode
must cover to be retianed {[}0.1{]}

\end{description}

\sphinxstylestrong{Type:} float; \sphinxstylestrong{Default:} 0.1


\subsubsection{\sphinxstyleliteralintitle{-D/-{-}mindepth}}
\label{\detokenize{prog_desc:d-mindepth}}
\sphinxstylestrong{Description:} minimum number of alleles per site

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 4


\subsubsection{\sphinxstyleliteralintitle{-M/-{-}minsites}}
\label{\detokenize{prog_desc:m-minsites}}
\sphinxstylestrong{Description:} minimum number of sites

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 100


\subsubsection{\sphinxstyleliteralintitle{-R/-{-}raxmlopts}}
\label{\detokenize{prog_desc:r-raxmlopts}}
\sphinxstylestrong{Description:} specify additional RAxML arguments as a double-quotes encased string

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:}


\subsubsection{\sphinxstyleliteralintitle{-X/-{-}raxmlpath}}
\label{\detokenize{prog_desc:id98}}
\sphinxstylestrong{Description:} RAxML path for manual specification.

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} raxml


\section{vcf2mvf}
\label{\detokenize{prog_desc:vcf2mvf}}

\subsection{Description}
\label{\detokenize{prog_desc:id99}}
MVFtools: Multisample Variant Format Toolkit
James B. Pease and Ben K. Rosenzweig
\sphinxurl{http://www.github.org/jbpease/mvftools}


\subsection{Parameters}
\label{\detokenize{prog_desc:id100}}

\subsubsection{\sphinxstyleliteralintitle{-h/-{-}help}}
\label{\detokenize{prog_desc:id101}}
\sphinxstylestrong{Description:} show this help message and exit

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}out} (required)}
\label{\detokenize{prog_desc:out-required}}
\sphinxstylestrong{Description:} output MVF file

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}vcf} (required)}
\label{\detokenize{prog_desc:vcf-required}}
\sphinxstylestrong{Description:} input VCF file

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}allelesfrom}}
\label{\detokenize{prog_desc:allelesfrom}}\begin{description}
\item[{\sphinxstylestrong{Description:} get additional alignment columns}] \leavevmode
from INFO fields (:-separated)

\end{description}

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}contigids}}
\label{\detokenize{prog_desc:contigids}}\begin{description}
\item[{\sphinxstylestrong{Description:} manually specify one or more contig ids}] \leavevmode
as ID;VCFLABE;MVFLABEL, note that
VCFLABEL must match EXACTLY the contig string
labels in the VCF file

\end{description}

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}fieldsep}}
\label{\detokenize{prog_desc:fieldsep}}
\sphinxstylestrong{Description:} VCF field separator (default=’TAB’)

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} TAB

\sphinxstylestrong{Choices:} {[}‘TAB’, ‘SPACE’, ‘DBLSPACE’, ‘COMMA’, ‘MIXED’{]}


\subsubsection{\sphinxstyleliteralintitle{-{-}linebuffer}}
\label{\detokenize{prog_desc:linebuffer}}
\sphinxstylestrong{Description:} number of lines to hold in read/write buffer

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 100000


\subsubsection{\sphinxstyleliteralintitle{-{-}lowdepth}}
\label{\detokenize{prog_desc:lowdepth}}
\sphinxstylestrong{Description:} below this read depth coverage, convert to lower case set to 0 to disable

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 3


\subsubsection{\sphinxstyleliteralintitle{-{-}lowqual}}
\label{\detokenize{prog_desc:lowqual}}\begin{description}
\item[{\sphinxstylestrong{Description:} below this quality convert to lower case}] \leavevmode
set to 0 to disable

\end{description}

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 20


\subsubsection{\sphinxstyleliteralintitle{-{-}maskdepth}}
\label{\detokenize{prog_desc:maskdepth}}
\sphinxstylestrong{Description:} below this read depth mask with N/n

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 1


\subsubsection{\sphinxstyleliteralintitle{-{-}maskqual}}
\label{\detokenize{prog_desc:maskqual}}\begin{description}
\item[{\sphinxstylestrong{Description:} low quality cutoff, bases replaced by N/-}] \leavevmode
set to 0 to disable

\end{description}

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 3


\subsubsection{\sphinxstyleliteralintitle{-{-}no\_autoindex}}
\label{\detokenize{prog_desc:no-autoindex}}
\sphinxstylestrong{Description:} do not automatically index contigs from the VCF

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}outflavor}}
\label{\detokenize{prog_desc:outflavor}}
\sphinxstylestrong{Description:} choose output MVF flavor to include quality scores and/or indels

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} dna

\sphinxstylestrong{Choices:} {[}‘dna’, ‘dnaqual’, ‘dnaqual-indel’, ‘dna-indel’{]}


\subsubsection{\sphinxstyleliteralintitle{-{-}overwrite}}
\label{\detokenize{prog_desc:id102}}
\sphinxstylestrong{Description:} USE WITH CAUTION: force overwrite of outputs

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}qual}}
\label{\detokenize{prog_desc:qual}}
\sphinxstylestrong{Description:} Include Phred genotype quality (GQ) scores

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}reflabel}}
\label{\detokenize{prog_desc:reflabel}}
\sphinxstylestrong{Description:} label for reference sample (default=’REF’)

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} REF


\subsubsection{\sphinxstyleliteralintitle{-{-}samplereplace}}
\label{\detokenize{prog_desc:samplereplace}}\begin{description}
\item[{\sphinxstylestrong{Description:} one or more \sphinxurl{TAG:NEWLABEL} or TAG, items,}] \leavevmode
if TAG found in sample label, replace with
NEW (or TAG if NEW not specified)
NEW and TAG must each be unique

\end{description}

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\chapter{Frequently Asked Questions}
\label{\detokenize{faq:frequently-asked-questions}}\label{\detokenize{faq::doc}}
See also our forum at: \sphinxurl{https://groups.google.com/forum/\#!forum/mvftools}

Coming soon.


\chapter{Version History}
\label{\detokenize{version::doc}}\label{\detokenize{version:version-history}}
\sphinxstylestrong{2017-06-25}

\sphinxstyleemphasis{Major Upgrade}: Full manual documentation added, standardization and cleanup of paramaters and upgrades and bugfixes throughout.

\sphinxstylestrong{2017-05-18}

Fixes to VCF conversion for compatibility

\sphinxstylestrong{2017-04-10}

Added MVF-to-Phylip output conversion \sphinxcode{mvf2phy}

\sphinxstylestrong{2017-03-25}

Multiple bug fixes, merged and removed the development instance

\sphinxstylestrong{2016-02-15}

Fix to vcf2mvf for VCF with truncated entries

\sphinxstylestrong{2016-10-25}

Efficiency upgrades for mvfbase entry iteration.

\sphinxstylestrong{2016-09-10}

Minor fixes to gz reading and MVF chromoplot shading

\sphinxstylestrong{2016-08-02}

Python3 conversion, integrate analysis\_base

\sphinxstylestrong{2016-01-11}

fix for dna ambiguity characters

\sphinxstylestrong{2016-01-01}

Python3 compatiblity fix

\sphinxstylestrong{2015-12-31}

Header changesand cleanup

\sphinxstylestrong{2015-12-15}

Python3 compatibilty fix

\sphinxstylestrong{2015-09-04}

Small style fixes

\sphinxstylestrong{2015-06-09}

MVF1.2.1 upgrade

\sphinxstylestrong{2015-02-26}

Efficiency upgrades for iterators

\sphinxstylestrong{2015-02-01}

First Public Release


\chapter{License}
\label{\detokenize{version:license}}
MVFtools is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
MVFtools is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
You should have received a copy of the GNU General Public License
along with MVFtools.  If not, see \textless{}\sphinxurl{http://www.gnu.org/licenses/}\textgreater{}.


\chapter{Indices and tables}
\label{\detokenize{index:indices-and-tables}}\begin{itemize}
\item {} 
\DUrole{xref,std,std-ref}{genindex}

\item {} 
\DUrole{xref,std,std-ref}{modindex}

\item {} 
\DUrole{xref,std,std-ref}{search}

\end{itemize}



\renewcommand{\indexname}{Index}
\printindex
\end{document}